### ISSUE TITLE
About GraphSAGE sampling on weighted graph

### ISSUE THREAD

**Issue by jindeok** [rusty1s/pytorch_geometric#1961]
## ❓ Questions & Help
Hello.
I really appreciate for sharing implementation examples of GNN, However, I have a 1 short question.

I'm wondering what happens when I put weighted graph into GraphSAGE example.
Does NeighborhoodSampling consider It as unweighted(binary edge) graph?
or weight of edges affect the sampling process?

**Comment by rusty1s** [comment]
Currently, `SAGEConv` does not support weightes graphs, but `GraphConv` does (which is quite similar). Note that you need to pas both `edge_index` and `edge_weight` to the GNN op.

**Comment by 1byxero** [comment]
Can we use sparse matrix for the edge weights?

**Comment by rusty1s** [comment]
Do you mean using the `SparseTensor` class? That is possible by passing edge weights to the `value` argument:
```python
adj_t = SparseTensor(row=row, col=col, value=edge_weight, sparse_sizes=(N, N))
```

**Comment by 1byxero** [comment]
This means we need one dense tensor storing edge weights initially. Right?

I don't want to do that as that dense tensor is taking up too much GPU memory.
Is there any workaround?

**Comment by jindeok** [comment]
Thank you for your reply.!! It was very helpful.


Can I ask one more question?
I'm also wonder whether I can apply `NeighborSampler` function on **weighted graph** if I want to construct training set from sampling.

**Comment by rusty1s** [comment]
`NeighborSampler` returns a `e_id` tensor, which can be used to index the original edge weights:
```python
loader = NeighborSampler(edge_index, ...)
for batch_size, n_id, adjs:
    for edge_index, e_id, size in adjs:
        sampled_edge_weight = edge_weight[e_id]
```

**Comment by jindeok** [comment]
You mean `edge_weight = data.edge_attr` in here??

Actually I build weighted graph upon **networkx graph**
than I convert it to torch_geometric data using `from_networkx()` function 
and I think that this converting method do not support to convert **edge attribute** from networkx. 
(Is only structural information of graph converted?)

In short, Now I am stuck in converting networkx weighted graph into torch_geometric weighted graph data..
Can you give me some tips on it?? 

Thanks for reading it!

**Comment by rusty1s** [comment]
Can you give me a short example to illustrate this issue?

**Comment by jindeok** [comment]
Im sorry I found that `from_networkx()`method also convert edge weight.
anyway,

my work example]

*  I want to aggregate information from sampled neighbor batch with weighted graph *
`
# In graphsage_unsupervised example code, I just replaced SAGEConv into GraphConv here
class myGNN(nn.Module):
    def __init__(self, in_channels, hidden_channels, num_layers):
        super(myGNN, self).__init__()
        self.num_layers = num_layers
        self.convs = nn.ModuleList()
        for i in range(num_layers):
            in_channels = in_channels if i == 0 else hidden_channels
            self.convs.append(**GraphConv**(in_channels, hidden_channels, aggr = 'mean'))

**# how to design forward method to input weight of edges?**
    def forward(self, x, adjs):
        for i, (edge_index, _, size) in enumerate(adjs):
            x_target = x[:size[1]]  # Target nodes are always placed first.
            x = self.convs[i]((x, x_target), edge_index) # ADD edge weight  !!
            if i != self.num_layers - 1:
                x = x.relu()
                x = F.dropout(x, p=0.5, training=self.training)
        return x

    def full_forward(self, x, edge_index):
        for i, conv in enumerate(self.convs):
            x = conv(x, edge_index) # ADD edge weight  !!
            if i != self.num_layers - 1:
                x = x.relu()
                x = F.dropout(x, p=0.5, training=self.training)
        return x


data = from_networkx(G)  # G: Weighted Graph that constructed in networkx
train_loader = NeighborSampler(data.edge_index, sizes=[10, 10], batch_size=256,
                               shuffle=True, num_nodes=data.num_nodes)

model.train()

total_loss = 0
for batch_size, n_id, adjs in train_loader:
    **for edge_index, e_id, size in adjs:
        sampled_edge_weight = edge_weight[e_id]**        #I dont know how to deal with it ,,
        adjs = [adj.to(device) for adj in adjs]
        optimizer.zero_grad()
    
        out = model(....)
        out, pos_out, neg_out = out.split(out.size(0) // 3, dim=0)
    
        pos_loss = F.logsigmoid((out * pos_out).sum(-1)).mean()
        neg_loss = F.logsigmoid(-(out * neg_out).sum(-1)).mean()
        loss = -pos_loss - neg_loss
        loss.backward()
        optimizer.step()

`

**Comment by rusty1s** [comment]
You need to index select the edge weights coming from `data.edge_weight`. The correct example would look similar to:
```python
def forward(self, x, adjs, edge_weight):
        for i, (edge_index, e_id, size) in enumerate(adjs):
            x_target = x[:size[1]]  # Target nodes are always placed first.
            x = self.convs[i]((x, x_target), edge_index, edge_weight[e_id]) 
            if i != self.num_layers - 1:
                x = x.relu()
                x = F.dropout(x, p=0.5, training=self.training)

for batch_size, n_id, adjs in train_loader:
    ...
    model(x[n_id], adjs, data.edge_weight)
```

**Comment by pintonos** [comment]
> Currently, `SAGEConv` does not support weightes graphs, but `GraphConv` does (which is quite similar). Note that you need to pas both `edge_index` and `edge_weight` to the GNN op.

For the inductive case it seems  `SAGEConv` performs better than  `GraphConv`. Is there any possibility to extend the `SAGEConv` with edge weights? And after reading the paper of  `GraphConv` I am not sure why it should be similar to  `SAGEConv`?

**Comment by rusty1s** [comment]
`GraphConv` is the same as `SAGEConv` in case you specify `aggr="mean"`. Let me know if that works for you.

**Comment by pintonos** [comment]
> `GraphConv` is the same as `SAGEConv` in case you specify `aggr="mean"`. Let me know if that works for you.

Works fine now, thanks! The documentation of `GraphConv` only defines its node-wise formulation. What would be the definition for the whole graph as in `GCNConv` for instance?

**Comment by rusty1s** [comment]
```
X @ W_1 + D^{-1} A X W_2
```

**Comment by kangzhang0709** [comment]
> You need to index select the edge weights coming from `data.edge_weight`. The correct example would look similar to:
> 
> ```python
> def forward(self, x, adjs, edge_weight):
>         for i, (edge_index, e_id, size) in enumerate(adjs):
>             x_target = x[:size[1]]  # Target nodes are always placed first.
>             x = self.convs[i]((x, x_target), edge_index, edge_weight[e_id]) 
>             if i != self.num_layers - 1:
>                 x = x.relu()
>                 x = F.dropout(x, p=0.5, training=self.training)
> 
> for batch_size, n_id, adjs in train_loader:
>     ...
>     model(x[n_id], adjs, data.edge_weight)
> ```

I have a question about,How to generate positive and negative samples by using `NeighborSampler` Sampling on weighted graph?

**Comment by rusty1s** [comment]
We do not have support for biased/weighted sampling of random walks yet, I am sorry.

**Comment by ZRH0308** [comment]
> > You need to index select the edge weights coming from `data.edge_weight`. The correct example would look similar to:
> > ```python
> > def forward(self, x, adjs, edge_weight):
> >         for i, (edge_index, e_id, size) in enumerate(adjs):
> >             x_target = x[:size[1]]  # Target nodes are always placed first.
> >             x = self.convs[i]((x, x_target), edge_index, edge_weight[e_id]) 
> >             if i != self.num_layers - 1:
> >                 x = x.relu()
> >                 x = F.dropout(x, p=0.5, training=self.training)
> > 
> > for batch_size, n_id, adjs in train_loader:
> >     ...
> >     model(x[n_id], adjs, data.edge_weight)
> > ```
> 
> I have a question about,How to generate positive and negative samples by using `NeighborSampler` Sampling on weighted graph?
> 
> in this code `pos_batch = random_walk(row, col, batch, walk_length=1,coalesced=True)[:, 1]`,how to randow walk according to the edge weight?

Hi,
I have the same need as you, how did you solve it in the end? 
My idea is that if I only need to take one step neighbor, can I directly select the TOP K% neighbor with the largest edge_weight in the current batch for each node as positive samples?   Do you think it will work?  @rusty1s

**Comment by ZRH0308** [comment]
> We do not have support for biased/weighted sampling of random walks yet, I am sorry.

Hi, @rusty1s 
Now is it possible to do weighted sampling of random walks? I need to sample positive neighbors based on the edge_weight.

If I just need to take an one step neighbor, can I directly select the TOP neighbor with the largest edge_weight in the current batch for each node as a positive sample? Do you think it will work?

**Comment by rusty1s** [comment]
There is an open PR for this out, see https://github.com/rusty1s/pytorch_cluster/pull/140. Maybe you can check it out to see if it fits your need.

**Comment by chirag-sj** [comment]
> `GraphConv` is the same as `SAGEConv` in case you specify `aggr="mean"`. Let me know if that works for you.

**Comment by MaryamHamad** [comment]
Hello @rusty1s , I would like to ask what is the default number of nodes that GraphSAGE uses to sample the neighbour nodes (if not specified) ??
I read about GraphSage its main idea is to apply sampling for the neighbouring nodes (randomly ), can you explain please what is the number of ratio of nodes are used?
